core/src/main/java/tachyon/worker/hierarchy/StorageDir.java

```diff
@@ -105,63 +115,125 @@
    * @param blockId Id of the block
    */
   public void accessBlock(long blockId) {
-    mLastBlockAccessTimeMs.put(blockId, System.currentTimeMillis());
+    if (containsBlock(blockId)) {
+      mLastBlockAccessTimeMs.put(blockId, System.currentTimeMillis());
+    }
   }
 
   /**
-   * Add information of a block in current StorageDir
+   * Adds a block into the dir.
+   * 
+   * @param blockId the Id of the block
+   * @param sizeBytes the size of the block in bytes
+   * @param report need to be reported during heartbeat with master
+   */
+  private void addBlock(long blockId, long sizeBytes, boolean report) {
+    addBlock(blockId, sizeBytes, System.currentTimeMillis(), report);
+  }
+
+  /**
+   * Adds a block into the dir.
    * 
    * @param blockId Id of the block
    * @param sizeBytes size of the block in bytes
+   * @param accessTimeMs access time of the block in millisecond.
+   * @param report whether need to be reported During heart beat with master
    */
-  private void addBlockId(long blockId, long sizeBytes) {
-    accessBlock(blockId);
+  private void addBlock(long blockId, long sizeBytes, long accessTimeMs, boolean report) {
+    mLastBlockAccessTimeMs.put(blockId, accessTimeMs);
+    if (mBlockSizes.containsKey(blockId)) {
+      mSpaceCounter.returnUsedBytes(mBlockSizes.remove(blockId));
+    }
     mBlockSizes.put(blockId, sizeBytes);
+    if (report) {
+      mAddedBlockIdList.add(blockId);
+    }
   }
 
   /**
    * Move the cached block file from user temporary directory to data directory
    * 
-   * @param userId Id of the user
-   * @param blockId Id of the block
+   * @param userId the id of the user
+   * @param blockId the id of the block
    * @return true if success, false otherwise
    * @throws IOException
    */
   public boolean cacheBlock(long userId, long blockId) throws IOException {
     String srcPath = getUserTempFilePath(userId, blockId);
     String dstPath = getBlockFilePath(blockId);
+    Long allocatedBytes = mTempBlockAllocatedBytes.remove(new Pair<Long, Long>(userId, blockId));
+
+    if (!mFs.exists(srcPath) || allocatedBytes == null) {
+      cancelBlock(userId, blockId);
+      throw new IOException("Block file doesn't exist! blockId:" + blockId);
+    }
     long blockSize = mFs.getFileSize(srcPath);
-    boolean result = mFs.rename(srcPath, dstPath);
-    if (result) {
-      addBlockId(blockId, blockSize);
+    if (blockSize < 0) {
+      cancelBlock(userId, blockId);
+      throw new IOException("Negative size of block! blockId:" + blockId);
+    }
+    returnSpace(userId, allocatedBytes - blockSize);
```

Comment:
seems a bug. what if the file does not exist anymore and the ioexception was raised. this line won't be called in that case.